### PR TITLE
feat: implement proxy handler

### PR DESCRIPTION
**Pull request by abc3**
The current PR introduces a new handler that will replace session mode. It authenticates incoming connections via auth queries, establishes a TCP connection to the tenant's database, and then simply forwards data between the client and the database, functioning primarily as a proxy

- [x] cancel query
- [x] terminate auth_query connection
- [x] active once
- [x] application_name with "via supavisor"
- [x] handle client disconnect
- [x] handle DB disconnect
- [x] auth_query conn is terminated if the client is killed
- [x] telemetry
- [x] delete metrics  
- [x] tests

### CODE REVIEW

File: lib/supavisor/handlers/proxy/handler.ex [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+      connection_start: System.monotonic_time(),
+      log_level: nil,
+      version: Application.spec(:supavisor, :vsn),
+      auth: nil,
+      nonce: nil,
+      server_proof: nil,
+      parameter_status: %{},
+      app_name: nil,
+      peer_ip: Helpers.peer_ip(sock),
+      auth: %{},
```
Comment (hauleth) [inline lib/supavisor/handlers/proxy/handler.ex:74]: `auth` key is mentioned twice.

File: lib/supavisor/handlers/proxy/handler.ex [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+  ## Internal functions
+
+  @spec clean_up(map()) :: any()
+  defp clean_up(data) do
+    HH.sock_close(data.sock)
+    HH.sock_close(data.db_sock)
+
+    case Registry.lookup(Supavisor.Registry.TenantClients, data.id) do
+      clients when clients in [[{self(), []}], []] -> PromEx.remove_metrics(data.id)
```
Comment (hauleth) [inline lib/supavisor/handlers/proxy/handler.ex:151]: I have some issues with readability of this line. I do not know how to make it more readable, maybe something like:

```elixir
if List.first(lookup_result, {self(), []}) == {self(), []} do
  # …
```

Would be more readable?

File: lib/supavisor/handlers/proxy/handler.ex [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+  alias Supavisor.HandlerHelpers, as: HH
+
+  @sock_closed [:tcp_closed, :ssl_closed]
+  @proto [:tcp, :ssl]
+
+  @impl true
+  def start_link(ref, transport, opts) do
+    pid = :proc_lib.spawn_link(__MODULE__, :init, [ref, transport, opts])
+    {:ok, pid}
```
Comment (hauleth) [inline lib/supavisor/handlers/proxy/handler.ex:25]: I would use `:proc_lib.start_link/3` instead.

File: lib/supavisor/handlers/proxy/client.ex [hunk 4]
```diff
@@ -0,0 +1,662 @@
...
+
+            Telem.client_join(:fail, id)
+            {:stop, {:shutdown, :address_not_allowed}}
+
+          true ->
+            new_data = update_user_data(data, info, user, id, db_name, mode)
+
+            key = {:secrets, tenant_or_alias, user}
+
+            case auth_secrets(info, user, key, :timer.hours(24)) do
```
Comment (hauleth) [inline lib/supavisor/handlers/proxy/client.ex:175]: Deeply nested conditionals make it hard to follow the flow of the code.

File: lib/supavisor/handlers/proxy/client.ex [hunk 5]
```diff
@@ -0,0 +1,662 @@
...
+      {:error, reason} ->
+        Logger.error(
+          "ProxyClient: User not found: #{inspect(reason)} #{inspect({type, user, tenant_or_alias})}"
+        )
+
+        :ok = HH.send_error(sock, "XX000", "Tenant or user not found")
+        Telem.client_join(:fail, data.id)
+        {:stop, {:shutdown, :user_not_found}}
+    end
+  end
```
Comment (hauleth) [inline lib/supavisor/handlers/proxy/client.ex:202]: This function is quite big.

File: lib/supavisor/handlers/proxy/client.ex [hunk 6]
```diff
@@ -0,0 +1,662 @@
...
+            do: Server.error_message("XX000", reason),
+            else: Server.exchange_message(:final, "e=#{reason}")
+
+        key = {:secrets_check, data.tenant, data.user}
+
+        if method != :password and reason == "Wrong password" and
+             Cachex.get(Supavisor.Cache, key) == {:ok, nil} do
+          case auth_secrets(info, data.user, key, 15_000) do
+            {:ok, {method2, secrets2}} = value ->
+              if method != method2 or Map.delete(secrets.(), :client_key) != secrets2.() do
```
Comment (hauleth) [inline lib/supavisor/handlers/proxy/client.ex:229]: Deeply nested conditionals again.

File: lib/supavisor/handlers/proxy/client.ex [hunk 7]
```diff
@@ -0,0 +1,662 @@
...
+      ip_ver
+    ]
+
+    case :gen_tcp.connect(~c"#{auth.host}", auth.port, sock_opts) do
+      {:ok, sock} ->
+        Logger.debug("ProxyClient: auth #{inspect(auth, pretty: true)}")
+
+        case ProxyDb.try_ssl_handshake({:gen_tcp, sock}, auth) do
+          {:ok, sock} ->
+            case ProxyDb.send_startup(sock, auth) do
```
Comment (hauleth) [inline lib/supavisor/handlers/proxy/client.ex:294]: Deeply nested conditional.

**Review by hauleth — APPROVED** [review]
Some compilation warnings to cleanup, and some readability issues that I have found.

**Review by hauleth — CHANGES_REQUESTED** [review]
I have clicked incorrectly previous time.

File: lib/supavisor/handlers/proxy/handler.ex (hunk 3 above)
Comment (abc3) [reply lib/supavisor/handlers/proxy/handler.ex:25]: Any reason for that? In the cowboy codebase, they mostly use spawn_link

File: lib/supavisor/handlers/proxy/handler.ex (hunk 2 above)
Comment (abc3) [reply lib/supavisor/handlers/proxy/handler.ex:151]: for me, this variant is harder to read :D

**Comment by abc3** [comment]
nested or long functions can be simplified before merging v2 into main, once the handler for transaction mode is ready

File: lib/supavisor/handlers/proxy/handler.ex (hunk 3 above)
Comment (hauleth) [reply lib/supavisor/handlers/proxy/handler.ex:25]: Somehow `start_link/3` is cleaner for me, as I can return error from spawned process (and also inform the parent when the handler is ready).

File: lib/supavisor/handlers/proxy/handler.ex (hunk 2 above)
Comment (hauleth) [reply lib/supavisor/handlers/proxy/handler.ex:150]: Alternatively we could expand the guard to full form `when clients == [{self(), []}] or clients == []`

Comment (abc3) [reply lib/supavisor/handlers/proxy/handler.ex:150]: yeah, that is good
